Fix performance regression with `spack mirror create --all`

alalazo opened the pull request:
fixes #31985 

This PR fixes the performance regression reported in #31985 and a few other issues found while refactoring the `spack mirror create` command (e.g. default directory not canonicalized, specs read from text file treated differently than specs from command line, allowing `-n` together with `--all`, etc.).

Modifications:
- [x] Do not require concretization for `spack mirror create --all`
- [x] Forbid using `--versions-per-spec` together with `--all`
- [x] Fixed a few issues when reading specs from input file (specs were not concretized, command would fail when trying to mirror dependencies)
- [x] Fix issue with default directory for `spack mirror create` not being canonicalized
- [x] Changed slightly the wording for reporting (it was mentioning "Successfully  created" even in presence of errors)
- [x] Fix issue with `colify` (was not called properly during error reporting)
- [x] Add more unit tests to poke `spack mirror create` 
- [x] Skip externals also when mirroring environments

alalazo committed:
Move error checking out of function

The function is meant to determine specs.

Error checking on arguments should be done as the first
thing in the main "mirror create" function, so that we
can assume arguments are sanitized when calling helper
functions.

outside, so that we can assume arguments are sanitized  then

alalazo committed:
Moved another args sanitization at the top of main function

alalazo committed:
Extracted function to retrieve the number of versions per spec

alalazo committed:
Moved another args sanitization at the top of main function

alalazo committed:
Forbid using -n and --all together

alalazo committed:
Fixed a dcstring, changed argument to reflect object type

alalazo committed:
Fix issue with --all

Per help --all implies fetching all the versions
of all the packages.

alalazo committed:
Fix an issue when reading specs from file

The specs were not concretized when read from file.

alalazo committed:
Refactored functions to create mirror from individual specs

alalazo committed:
Creating mirror with --all doesn't require concretization

Restore the previous behavior where creating a mirror with
--all didn't require concretizing specs.

Since due to this refactor spack.mirror._add_single_spec
does not exist anymore, move the regression test for 31736.

alalazo committed:
fixup

alalazo committed:
Add unit test on error conditions for mirror create

alalazo committed:
Fix issue with default directory not being canonicalized

alalazo committed:
Removed outdated portion of docstring

alalazo committed:
Add more unit tests to spack mirror create

alalazo committed:
Fix doc building

alalazo committed:
Extract a generic function to partition specs on a predicate

alalazo commented:
@Paul-Ferrell Can you test this branch on your use case? It should fix all the issues you've seen so far. Note though that here I forbid using `--all` together with `--versions-per-spec`, so you have to remove the latter option. Let me know if you still see issues.

alalazo commented:
For the record, I'll stop here with refactoring, but there are many more improvements that can be done in future PRs related to mirrors:
- The classes representing a mirror are complex and scarcely related to each other. I was surprised for instance to see that a `Mirror` has no method to push or pull something, since that is done using a cache which is a totally separate object. Even mirror statistics are a separate object from both caches and mirrors. I think there's room here to have a more coherent design.
- Stages and fetchers are too intertwined and require a package object to be instantiated cleanly. Making them more loosely coupled can enable using only fetchers in mirrors, which would make it easier to fetch tarballs in parallel.

Paul-Ferrell commented:
Much better. I am getting this error though: 
```
lib/spack/spack/error.py:54 ==> [2022-08-09-06:33:20.936530] Error: Package 'base' not found.
Traceback (most recent call last):
  File "/storage/spack/spack/lib/spack/spack/repo.py", line 1306, in get_pkg_class
    module = importlib.import_module(fullname)
  File "/usr/lib64/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'spack.pkg.builtin.base'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/storage/spack/spack/lib/spack/spack/main.py", line 961, in main
    return _main(argv)
  File "/storage/spack/spack/lib/spack/spack/main.py", line 916, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/storage/spack/spack/lib/spack/spack/main.py", line 944, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/storage/spack/spack/lib/spack/spack/main.py", line 608, in _invoke_command
    return_val = command(parser, args)
  File "/storage/spack/spack/lib/spack/spack/cmd/mirror.py", line 492, in mirror
    action[args.mirror_command](args)
  File "/storage/spack/spack/lib/spack/spack/cmd/mirror.py", line 425, in mirror_create
    create_mirror_for_all_specs(args)
  File "/storage/spack/spack/lib/spack/spack/cmd/mirror.py", line 442, in create_mirror_for_all_specs
    pkg_cls = spack.repo.path.get_pkg_class(candidate.name)
  File "/storage/spack/spack/lib/spack/spack/repo.py", line 962, in get_pkg_class
    return self.repo_for_pkg(pkg_name).get_pkg_class(pkg_name)
  File "/storage/spack/spack/lib/spack/spack/repo.py", line 1308, in get_pkg_class
    raise UnknownPackageError(pkg_name)
spack.repo.UnknownPackageError: Package 'base' not found.
```

alalazo commented:
@Paul-Ferrell Can you give me more context (which command you were executing, configuration if relevant)? What I see is that `spack` fails at finding the `base` package in the builtin repository, which is correct. What I don't see is why it is looking for that.

alalazo committed:
Racket packages shouldn't override the "name" attribute

alalazo commented:
@Paul-Ferrell Can you re-try? We had a few packages overriding an attribute that was meant for internal use.

Paul-Ferrell commented:
Ok, I think this one's a winner. After a couple minutes we're already up to the letter 'G' without any issues.

Paul-Ferrell commented:
I was checking on a different problem (more on that later) and ran into this when mirroring a single package:
```
  File "/home/pflarr/work_space/spack-mirror/bin/spack", line 100, in <module>
    sys.exit(spack.main.main())
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/main.py", line 961, in main
    return _main(argv)
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/main.py", line 916, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/main.py", line 944, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/main.py", line 608, in _invoke_command
    return_val = command(parser, args)
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/cmd/mirror.py", line 492, in mirror
    action[args.mirror_command](args)
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/cmd/mirror.py", line 432, in mirror_create
    create_mirror_for_individual_specs(args)
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/cmd/mirror.py", line 459, in create_mirror_for_individual_specs
    create_mirror_from_concrete_specs(mirror_specs, args)
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/cmd/mirror.py", line 365, in create_mirror_from_concrete_specs
    directory, mirror_specs, args.skip_unstable_versions
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/mirror.py", line 508, in create
    create_mirror_from_package_object(spec.package, mirror_cache, mirror_stats)
  File "/home/pflarr/work_space/spack-mirror/lib/spack/spack/spec.py", line 1521, in package
    assert self.concrete, "Spec.package can only be called on concrete specs"
AssertionError: Spec.package can only be called on concrete specs
```

Paul-Ferrell commented:
The other problem I ran into is throttling from gitlab.com. I think it will be fine over multiple runs, as the packages that got missed should just get picked up later.

alalazo committed:
Fix the case of using -n with a single spec

alalazo committed:
Reuse mock mirror args

alalazo commented:
The last commit should fix https://github.com/spack/spack/pull/32005#issuecomment-1209458406 It won't be as fast as it could, since it needs to concretize, but I'd rather try to speed that up in a following PR.

alalazo committed:
Change specs to something the old concretizer can handle

haampie commented on lib/spack/spack/spec.py:
```diff
@@ -5364,6 +5364,26 @@ def save_dependency_specfiles(
             fd.write(dep_spec.to_json(hash=ht.dag_hash))
 
 
+def partition_spec_list(input_specs, predicate_fn):
```
Can you move this elsewhere? It's just a generic stable partition

alalazo replied:
Moved to `llnl.util.lang` and renamed input args + reworked docstring

haampie commented on lib/spack/spack/spec.py:
```diff
@@ -5364,6 +5364,26 @@ def save_dependency_specfiles(
+    for item in input_specs:
+        if predicate_fn(item):
+            true_items.append(item)
+            continue
+        false_items.append(item)
```
more fancy than necessary imho ;p 
```suggestion
        else:
           false_items.append(item)
```

alalazo committed:
Move stable partition to llnl.util.lang

alalazo committed:
Remove types from docstring, since it's generic

scheibelp commented on lib/spack/llnl/util/lang.py:
```diff
@@ -977,6 +977,27 @@ def enum(**kwargs):
+    Args:
+        input_iterable: input iterable to be partitioned.
+        predicate_fn: predicate function accepting an iterable item
+            as argument.
```
(minor) might be worth mentioning that we `bool()` the result of `predicate_fn` (i.e. we don't strictly require `True/False` and will e.g. interpret values like `None` as `False`).

alalazo replied:
I don't think we should document that, nor we should complicate the implementation. So I added static checks with type hints to say what we expect.

scheibelp replied:
> So I added static checks with type hints to say what we expect.

That seems fine to me, but I'm curious before this thread gets collapsed: do we use this approach anywhere else? I had to look at https://github.com/spack/spack/pull/32005/commits/ecf225f167ee203355e885e37044d454e7d93fa5 to figure out what you meant. Usually we specify constraints on arguments in the docstring (or enforce with assert/raise in the method body).

alalazo replied:
I think this should be caught by `mypy` during style checks in CI. We have similar annotations in some class members here and there.

scheibelp commented on lib/spack/spack/build_systems/racket.py:
```diff
@@ -40,13 +40,13 @@ class RacketPackage(PackageBase):
     pkgs = False
     subdirectory = None  # type: Optional[str]
-    name = None  # type: Optional[str]
+    racket_name = None  # type: Optional[str]
```
(minor) Is it worth adding some package verification step that ensures that "reserved" members are not overridden? Not in this PR but later on (we could add a TODO here though).

alalazo replied:
Sure, I'll try to setup a different PR to do that.

scheibelp commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -231,142 +234,225 @@ def mirror_list(args):
+    with open(filename, "r") as f:
+        specs_in_file = f.readlines()
+        specs_in_file = [s.strip() for s in specs_in_file]
+    return spack.cmd.parse_specs(" ".join(specs_in_file), concretize=concretize)
```
(question) is there a benefit to passing all of these to the same call?

alalazo replied:
Yes, it ensures that specs that are concretized are always treated in the same way, e.g. if they fail the concretization step they emit the same error with the same message. This, in the end, gives a more consistent behavior of the command.

scheibelp replied:
Got it, my original concern was that it would concretize all the specs together (which would not be desirable for mirrors). AFAIK our API doesn't make any guarantees one way or the other (e.g. I see no unit tests enforcing this or docstrings mentioning this behavior explicitly).

scheibelp commented on lib/spack/spack/test/cmd/mirror.py:
```diff
@@ -288,3 +295,115 @@ def test_mirror_destroy(
+            # Passed -n along with --all
+            (
+                {"specs": None, "file": None, "all": True, "versions_per_spec": 2},
+                "cannot specify '--versions_per-spec'",
```
Why is `-n -a` not allowed? I have found this useful in the past (e.g. it is convenient to make a mirror with the latest version of each package e.g. with `-a -n 1`)

alalazo replied:
Because I took the help message as the declared intent of what the function should do:
```console
  -a, --all             mirror all versions of all packages in Spack, or all packages in the current environment if there is an active environment (this requires significant time and space)
```
If we want to have something that collects only the latest version, can we add an argument like `--all-latest` to do that?

alalazo replied:
Also, can we do this in a follow up PR?

scheibelp replied:
Yes. If I'm the only one who's ever used it, I don't mind if it takes a while to come back, it doesn't look hard to reintroduce (i.e. it doesn't look like any of the changes in this PR make it difficult to extract a limited number of versions for each package).

scheibelp commented on lib/spack/spack/test/cmd/mirror.py:
```diff
@@ -288,3 +295,115 @@ def test_mirror_destroy(
+        "cli_args,expected_end",
+        [
+            ({"directory": None}, os.path.join("var", "spack", "cache")),
+            ({"directory": os.path.join("foo", "bar")}, os.path.join("foo", "bar")),
```
For a relative path, I think this produces `file://foo/bar`: is that valid?

Also I think this should use `mock_config` (since the fallback behavior depends on the config).

alalazo replied:
See 2d48799

scheibelp commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -231,142 +234,225 @@ def mirror_list(args):
+    process_mirror_stats(*mirror_stats.stats())
+
+
+def create_mirror_for_all_specs_inside_environment(args):
```
Would it be possible to resolve `args` into the actual arguments used by each function at this level?

For example I think the only thing this needs is the excluded specs.

(When I say "at this level" I mean for this function as well as `create_mirror_for_individual_specs`, `create_mirror_for_all_specs`, etc.)

alalazo replied:
> For example I think the only thing this needs is the excluded specs

It needs also the `directory` where to place the mirror and whether or not to skip unstable versions. 

> Would it be possible to resolve args into the actual arguments used by each function at this level?

Done, though I have to say I slightly prefer the previous version of the code.[^1]  

[^1]: I prefer the previous version since the main `mirror_create` function was just doing error checking and then forwarding to other function if `args` passed all checks - so a terser view of the overall behavior without further details expanded. It's a minor style preference though.

scheibelp replied:
Thanks I appreciate this change (I often get confused when `args` from commands is passed transitively, since for me the method signature helps me understand at-a-glance what the function is "planning" to do, and when the parameters are grouped into a single opaque object I have to read the whole function and possibly function subcalls to understand it).

scheibelp reviewed (changes requested):
I have a couple requests and a couple questions. The largest requests (which I think are small) are:

* Process `args` into actual named variables before calling methods like `create_mirror_for_individual_specs`
* Allow `-a -n 1`

(I have specific comments on these with more detail)

I don't expect either would be time-consuming to address, but if I didn't estimate that right, let me know.

alalazo committed:
Use mock configuration, ensure path is absolute

alalazo committed:
local_mirror_url_from_user: renamed function, expanded arguments

alalazo committed:
create_mirror_for_individual_specs: expanded arguments

alalazo committed:
extend_with_additional_versions: extracted function

Also, this modification is now done *before* extending
with dependencies. The rationale is that if users wants
2 versions of "hdf5~mpi" in mirror it doesn't mean to get
2 versions for each dependency.

alalazo committed:
create_mirror_for_individual_specs: expanded arguments

alalazo committed:
create_mirror_for_all_specs: expanded arguments

alalazo committed:
create_mirror_for_all_specs_inside_environment: expanded arguments

alalazo committed:
stable_partition: add type hints

alalazo commented:
@scheibelp Ready for another pass

scheibelp reviewed (approved):
I have a very minor concern related to future proofing at https://github.com/spack/spack/pull/32005#discussion_r944010626 but this seems good

scheibelp commented:
Thanks!
